network/dag/state.go

```diff
@@ -24,73 +24,65 @@ import (
 	"github.com/nuts-foundation/nuts-node/core"
 	"github.com/nuts-foundation/nuts-node/crypto/hash"
 	"github.com/nuts-foundation/nuts-node/network/dag/tree"
 	"github.com/nuts-foundation/nuts-node/network/log"
-	"github.com/nuts-foundation/nuts-node/network/storage"
-	"go.etcd.io/bbolt"
 )
 
 const (
-	// boltDBFileMode holds the Unix file mode the created BBolt database files will have.
-	boltDBFileMode = 0600
 	// PageSize specifies the Lamport Clock range over which data is summarized and is used in set reconciliation.
 	PageSize = uint32(512)
 	// IbltNumBuckets is the number of buckets in the IBLT used in set reconciliation.
 	IbltNumBuckets = 1024
+	xorShelf       = "xorBucket"
+	ibltShelf      = "ibltBucket"
 )
 
 // State has references to the DAG and the payload store.
 type state struct {
-	db                               *bbolt.DB
-	graph                            *bboltDAG
+	db                               stoabs.KVStore
+	graph                            *dag
 	payloadStore                     PayloadStore
 	transactionalObservers           []Observer
 	nonTransactionalObservers        []Observer
 	transactionalPayloadObservers    []PayloadObserver
 	nonTransactionalPayloadObservers []PayloadObserver
 	txVerifiers                      []Verifier
-	xorTree                          *bboltTree
-	ibltTree                         *bboltTree
+	xorTree                          *treeStore
+	ibltTree                         *treeStore
+	treeMux                          sync.Mutex
 }
 
 // NewState returns a new State. The State is used as entry point, it's methods will start transactions and will notify observers from within those transactions.
-func NewState(dataDir string, verifiers ...Verifier) (State, error) {
-	dbFile := path.Join(dataDir, "network", "data.db")
-	if err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm); err != nil {
-		return nil, fmt.Errorf("unable to create BBolt database: %w", err)
-	}
-
-	var bboltErr error
-	db, bboltErr := bbolt.Open(dbFile, boltDBFileMode, bbolt.DefaultOptions)
-	if bboltErr != nil {
-		return nil, fmt.Errorf("unable to create BBolt database: %w", bboltErr)
-	}
-
-	graph := newBBoltDAG(db)
+func NewState(db stoabs.KVStore, verifiers ...Verifier) (State, error) {
+	graph := newDAG(db)
 
-	payloadStore := NewBBoltPayloadStore(db)
+	payloadStore := NewPayloadStore()
 	newState := &state{
 		db:           db,
 		graph:        graph,
 		payloadStore: payloadStore,
 		txVerifiers:  verifiers,
 	}
 
-	xorTree := newBBoltTreeStore(db, "xorBucket", tree.New(tree.NewXor(), PageSize))
-	ibltTree := newBBoltTreeStore(db, "ibltBucket", tree.New(tree.NewIblt(IbltNumBuckets), PageSize))
-	newState.xorTree = xorTree
-	newState.ibltTree = ibltTree
-	newState.RegisterTransactionObserver(newState.treeObserver, true)
-
 	return newState, nil
 }
 
+func initShelfs(db stoabs.KVStore, shelfs ...string) error {
+	for _, shelf := range shelfs {
+		if err := db.WriteShelf(shelf, func(writer stoabs.Writer) error {
```

Comment:
This will probably won't do anything when using Redis

Reply:
The `stoabs.ReadTx` interface for `GetShelfReader(...)` states that if a shelf does not exist, the returned `stoabs.Reader == nil`. 

For bbolt `initShelfs` means that we don't have to check if the reader is nil every time. Redis might still fail depending on its implementation. The existence of a shelf in Redis could be derived from a partial key match, but for certain shelfs (connections/payload_jobs) existence is not well defined/transient. This is rather confusing. I'm not sure what the right way of dealing with this is.

Any suggestions?

Reply:
Mark current lines with a TODO.
Add an issue to go-stoabs for a feature request: support DB specific initialisation scripts.

Reply:
https://github.com/nuts-foundation/go-stoabs/issues/14

Reply:
Maybe it would be simpler to have `GetShelfReader` return a reader, even when the shelf doesn't exist. But all read operations will return a value that indicates the key doesn't exist.

The check for Redis to determine "does shelf exist" involves doing a wildcard match for the shelf and then count the results, which doesn't seem very efficient to me.

If the existence check is required, we can add it as separate function (`store.ShelfExists()`)?

Reply:
To me that would be a reasonable outcome of the issue mentioned above, but probably out of the scope of this PR?

Reply:
The empty reader would be a good solution. https://github.com/nuts-foundation/go-stoabs/issues/14 may be left open for future use.